src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc

```diff
@@ -43,158 +43,310 @@ using Topology = atlas::mesh::Nodes::Topology;
+  //      owned counterparts on the other side of the partition boundary.
+  //
+  // Global indices of cells are the same as the grid point indices. Global
+  // indices of nodes first count the owned nodes, then the type-A ghost points.
+  //
+  // Local indices of cells follows the order of the local grid points. Local
+  // indces of the nodes first count the owned nodes, followed by type-A ghost
+  // nodes, followed by type-B ghost nodes.
+  //
+  // There are several stages to the mesh generator:
+  //    1. Preamble.
+  //    2. Define global cell distribution.
+  //    3. Define global owned node and type-A ghost node distribution.
+  //    4. Locate and count local type-B ghost nodes.
+  //    5. Assign nodes to mesh.
+  //    6. Assign cells to mesh.
+  //    7. Finalise.
+
+  // ---------------------------------------------------------------------------
+  // 1. PREAMBLE
+  // ---------------------------------------------------------------------------
 
   // Get dimensions of grid
   const auto N      = csGrid.N();
   const auto nTiles = csGrid.GetNTiles();
 
-  ATLAS_TRACE("CubedSphereMeshGenerator::generate");
-  Log::debug() << "Number of cells per tile edge = " << std::to_string(N) << std::endl;
+  const auto nNodesUnique = nTiles * N * N + 2;
+  const auto nNodesAll    = nTiles * (N + 1) * (N + 1);
+  const auto nCells       = nTiles * N * N;
+
+  Log::debug() << "Number of cells per tile edge = "
+    << std::to_string(N) << std::endl;
+
+  // Define bad index values.
+  constexpr auto undefinedIdx = -1;
+  constexpr auto undefinedGlobalIdx = -1;
+
+  // Projection and tiles
+  const auto* const csProjection =
+    dynamic_cast<const projection::detail::CubedSphereProjectionBase*>(
+    csGrid.projection().get());
+  ATLAS_ASSERT(csProjection);
+  const auto csTiles = csProjection->getCubedSphereTiles();
+
+  // Get partition information.
+  const auto nParts =   mpi::comm().size();
+  const auto thisPart = mpi::comm().rank();
+
+  // Set some casting helper functions to avoid annoying warnings.
+  const auto st2idx = [](size_t i){return static_cast<idx_t>(i);};
+  const auto idx2st = [](idx_t i){return static_cast<size_t>(i);};
+
+  // Define an index counter.
+  const auto idxSum = [](const std::vector<idx_t>& idxCounts) -> idx_t {
+    return std::accumulate(idxCounts.begin(), idxCounts.end(), 0);};
+
+  // Helper functions to get node and cell idx from (t, j, i).
+  const auto getNodeIdx = [&](idx_t t, idx_t j, idx_t i){
+    // Adjust bounds.
+    t = std::max(std::min(t, nTiles - 1), 0);
+    j = std::max(std::min(j, N), 0);
+    i = std::max(std::min(i, N), 0);
+    return idx2st(t * (N + 1) * (N + 1) + j * (N + 1) + i);
+  };
+
+  const auto getCellIdx = [&](idx_t t, idx_t j, idx_t i){
+    // Adjust bounds.
+    t = std::max(std::min(t, nTiles - 1), 0);
+    j = std::max(std::min(j, N - 1), 0);
+    i = std::max(std::min(i, N - 1), 0);
+    return idx2st(t * N * N + j * N + i);
+  };
+
```

Comment:
All of the above are named lambdas, like getCellIdx, getNodeIdx, idxSum, etc. They could be simple functions instead. You don't need to change it here, but I still believe that the usage of these here is not conform to how lambdas are intended to be used in C++. Also, for people not very familiar with C++, reading this code might be more complicated than if you were using short functions.